src/common/utility/array-replace.ts

```diff
@@ -0,0 +1,90 @@
+import { isArray } from 'lodash';
+
+import isPlainObject from './is-plain-object';
+import objectMerge from './object-merge';
+import replace from './replace';
+
+export interface ArrayReplaceOptions {
+    matchObject(objectA: any, objectB: any): boolean;
+}
+
+export default function arrayReplace<T>(currentArray: T[] | undefined, newArray: T[], options?: ArrayReplaceOptions): T[];
+export default function arrayReplace<T>(currentArray: T[], newArray?: T[], options?: ArrayReplaceOptions): undefined;
+export default function arrayReplace<T>(currentArray?: T[], newArray?: T[], options?: ArrayReplaceOptions): T[] | undefined;
+export default function arrayReplace<T>(currentArray?: T[], newArray?: T[], options?: ArrayReplaceOptions): T[] | undefined {
+    const { matchObject = (a: any, b: any) => a.id === b.id } = options || {};
+
+    // Return the new array if the current array does not exist
+    if (!currentArray && newArray || !currentArray && !newArray) {
+        return newArray;
+    }
+
+    // If the new array does not exist, or the current array is strictly equal
+    // to the new array, exit early and return the current array so that the
+    // object reference doesn't change
+    if (!newArray || currentArray === newArray) {
+        return currentArray;
+    }
+
+    // Otherwise, try to replace the items of the current array with the new
+    // array. If the item from the two arrays are the same, keep the current
+    // one. Do it recursively until all arrays are replaced.
+    let sameAsCurrentCount = 0;
+    let sameAsNewCount = 0;
+
+    const countSameAsReplaced = <T>(replacedValue: T, currentValue: T, newValue: T): T => {
+        sameAsCurrentCount += replacedValue === currentValue ? 1 : 0;
+        sameAsNewCount += replacedValue === newValue ? 1 : 0;
+
+        return replacedValue;
```

Comment:
Maybe for keeping the logic slightly simpler you just need to keep the count for `newItemsCounts` ? If `newItemsCount = 0` you know you can return `currentArray`,
if `newItemsCount === newArray.length`, then `newArray`, otherwise, modified array?

Reply:
Thanks Luis, I'll give it a go. It seems I might be able to just use a single counter instead of two :)